tasks.py

```diff
@@ -0,0 +1,253 @@
+            shutil.rmtree(target)
+    for timestamp_file in [node_timestamp_file, node_lint_timestamp_file, python_lint_timestamp_file]:
+        timestamp_file.unlink(missing_ok=True)
+
+
+@task
+def protobuf(c):
+    if not python_protobuf_dir.exists():
+        python_protobuf_dir.mkdir()
+        c.run(f"touch {python_protobuf_dir / '__init__.py'}")
+    if not node_protobuf_dir.exists():
+        node_protobuf_dir.mkdir()
+    gen_timestamp_file = python_protobuf_dir / ".generated"
+    if _sources_changed(proto_sources, gen_timestamp_file):
+        _python_protobuf_gen(c)
+        _node_protobuf_gen(c)
+        gen_timestamp_file.touch()
+    else:
+        print("no changes in .proto files, skipping protobuf build")
+
+
+def _python_protobuf_gen(c):
+    c.run(
+        f"python -m grpc_tools.protoc -I protobuf --python_out=Browser/generated --grpc_python_out={python_protobuf_dir} --mypy_out={python_protobuf_dir} protobuf/*.proto"
+    )
+    genfile = python_protobuf_dir / "playwright_pb2_grpc.py"
+    content = (
+        open(genfile)
+        .read()
+        .replace(
+            "import playwright_pb2 as playwright__pb2",
+            "from Browser.generated import playwright_pb2 as playwright__pb2",
+        )
+    )
+    with open(genfile, "w") as outfile:
+        outfile.write(content)
+
+
+def _node_protobuf_gen(c):
+    protoc_plugin = root_dir / "node_modules" / ".bin" / "grpc_tools_node_protoc_plugin"
+    protoc_ts_plugin = root_dir / "node_modules" / ".bin" / "protoc-gen-ts"
+    c.run(
+        f"yarn run grpc_tools_node_protoc \
+		--js_out=import_style=commonjs,binary:{node_protobuf_dir} \
+		--grpc_out={node_protobuf_dir} \
+		--plugin=protoc-gen-grpc={protoc_plugin} \
+		-I ./protobuf \
+		protobuf/*.proto"
+    )
+    c.run(
+        f"yarn run grpc_tools_node_protoc \
+		--plugin=protoc-gen-ts={protoc_ts_plugin} \
+		--ts_out={node_protobuf_dir} \
+		-I ./protobuf \
+		protobuf/*.proto"
+    )
+
+
+@task(protobuf)
+def node_build(c):
+    if _sources_changed(node_dir.glob("**/*.ts"), node_timestamp_file):
+        c.run("yarn build")
+        node_timestamp_file.touch()
+    else:
+        print("no changes in .ts files, skipping node build")
+
+
+@task(protobuf, node_build)
+def build(c):
+    c.run("python -m Browser.gen_stub")
+
+
+def _sources_changed(source_files, timestamp_file):
+    if timestamp_file.exists():
+        last_built = timestamp_file.lstat().st_mtime
+        src_last_modified = [f.lstat().st_mtime for f in source_files]
+        return not all([last_built >= modified for modified in src_last_modified])
+    return True
+
+
+@task(build)
+def watch(c):
+    c.run("yarn watch")
+
+
+@task
+def utest(c):
+    status = pytest.main()
+    raise Exit()
+
+
+@task
+def utest_watch(c):
+    c.run("ptw --ignore ./node_modules --ignore ./.venv")
+
+
+@task
+def clean_atest(c):
+    if atest_output.exists():
+        shutil.rmtree(atest_output)
+
+
+@task(clean_atest)
+def atest(c):
```

Comment:
Should this depend on the build task (like the makefile used to) so it's always run on newest changes?

Reply:
Good point, I'll add it here